### ISSUE TITLE
Does ureq support HTTP/2 ?

### ISSUE THREAD

**Issue by gilescope** [algesten/ureq#540]
The PR for using ureq in git2-rs (and thus in cargo) was closed ( https://github.com/rust-lang/git2-rs/pull/625 ) due to ureq not supporting HTTP/2 at the time. I thought I'd check as time has passed and see if this is maybe something that is now supported (and could this unblock that PR)?

**Comment by algesten** [comment]
ureq doesn't support HTTP/2. Yet. The thing about http/2 is that it tries to optimize loading of web content by multiplexing several simultaneous requests into one TCP socket.

ureq is single threaded and currently blocking, which means there isn't a great fit for how to juggle multiple simultaneous stream of data.

We could of course "speak" http/2, but only do one stream at a time, but the question is what is gained by that.

We could try and model some new API so that both making the request and reading the response is multiplexed by having the user of the API spin up a thread for each stream.

Or we could make (another) async library – however that space is kinda filled. ureq found  niche of "sync" where there are less alternatives.

-----------

To sum it up: it's not entirely obvious what the gain http/2 is for a sync, single threaded library like ureq.

Potential gain would require significant new API and it's not clear this is where we should go.

Obviously happy to discuss it further because there might be use cases I don't see.

**Comment by adlerd** [comment]
This makes sense with the goal of simplicity. Can you please indicate somewhere in the docs that ureq is an HTTP/1.1 client though? It took way too long for me to figure this out as a potential user. (But I'll probably end up using ureq anyways because I agree that simple is often better!) For example, under Features: "HTTP/1.1 (compatible with almost every server out there)".

Anecdotally, probably 98% of my outgoing web traffic from everyday browser use is HTTP/2 or HTTP/3, so it's not very clear when saying "HTTP" nowadays that you mean 1.1.

**Comment by lolgesten** [comment]
@adlerd agree we should bring this out in the docs.

> "HTTP/1.1 (compatible with almost every server out there)"

Just curious about the wording, have you encountered HTTP/2 only servers? So far I don't think I have.

**Comment by algesten** [comment]
ureq is likely to be http/1.1 only forever.
